cmd/ci-secret-bootstrap/main.go

```diff
@@ -0,0 +1,285 @@
+				kc, ok := kubeConfigs[secretContext.Cluster]
+				if !ok {
+					return fmt.Errorf("config[%d].to[%d]: failed to find cluster context %q in the kubeconfig", i, j, secretContext.Cluster)
+				}
+				client, err := coreclientset.NewForConfig(&kc)
+				if err != nil {
+					return err
+				}
+				o.secretsGetters[secretContext.Cluster] = client
+			}
+		}
+	}
+
+	return o.validateCompletedOptions()
+}
+
+func (o *options) validateCompletedOptions() error {
+	if o.bwPassword == "" {
+		return fmt.Errorf("--bw-password-file was empty")
+	}
+	if len(o.config) == 0 {
+		return fmt.Errorf("--config was empty")
+	}
+	toMap := map[string]map[string]string{}
+	for i, secretConfig := range o.config {
+		if len(secretConfig.From) == 0 {
+			return fmt.Errorf("config[%d].from is empty", i)
+		}
+		if len(secretConfig.To) == 0 {
+			return fmt.Errorf("config[%d].to is empty", i)
+		}
+		for key, bwContext := range secretConfig.From {
+			if key == "" {
+				return fmt.Errorf("config[%d].from: empty key is not allowed", i)
+			}
+			if bwContext.BWItem == "" {
+				return fmt.Errorf("config[%d].from[%s]: empty value is not allowed", i, key)
+			}
+			if bwContext.Field == "" && bwContext.Attachment == "" {
+				return fmt.Errorf("config[%d].from[%s]: either field or attachment needs to be non-empty", i, key)
+			}
+			if bwContext.Field != "" && bwContext.Attachment != "" {
+				return fmt.Errorf("config[%d].from[%s]: cannot use both field and attachment", i, key)
+			}
+		}
+		for j, secretContext := range secretConfig.To {
+			if secretContext.Cluster == "" {
+				return fmt.Errorf("config[%d].to[%d].cluster: empty value is not allowed", i, j)
+			}
+			if secretContext.Namespace == "" {
+				return fmt.Errorf("config[%d].to[%d].namespace: empty value is not allowed", i, j)
+			}
+			if secretContext.Name == "" {
+				return fmt.Errorf("config[%d].to[%d].name: empty value is not allowed", i, j)
+			}
+
+			if toMap[secretContext.Cluster] == nil {
+				toMap[secretContext.Cluster] = map[string]string{secretContext.Namespace: secretContext.Name}
+			} else if toMap[secretContext.Cluster][secretContext.Namespace] != secretContext.Name {
+				toMap[secretContext.Cluster][secretContext.Namespace] = secretContext.Name
+			} else {
+				return fmt.Errorf("config[%d].to[%d]: secret %v showed more than once in the config", i, j, secretContext)
+			}
+		}
+	}
+	return nil
+}
+
+type bitWardenContext struct {
+	BWItem     string `json:"bw_item"`
+	Field      string `json:"field,omitempty"`
+	Attachment string `json:"attachment,omitempty"`
+}
+
+type secretContext struct {
+	Cluster   string `json:"cluster"`
+	Namespace string `json:"namespace"`
+	Name      string `json:"name"`
+}
+
+type secretConfig struct {
+	From map[string]bitWardenContext `json:"from"`
+	To   []secretContext             `json:"to"`
+}
+
+func constructSecrets(config []secretConfig, bwClient bitwarden.Client) (map[string][]*coreapi.Secret, error) {
```

Comment:
@stevekuznetsov 

The reason that the return is `map[string][]*coreapi.Secret` instead of `[]*coreapi.Secret` is that I need a place to save the cluster that this secret belongs to.

Just found an alternative
https://github.com/kubernetes/apimachinery/blob/44bd77c24ef93cd3a5eb6fef64e514025d10d44e/pkg/apis/meta/v1/types.go#L266-L270

I think we could use it. 
What do you think?

Reply:
Map is fine